### ISSUE TITLE
Cirrus CI jobs don't actually merge `main`

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#19375]
See [this log](https://cirrus-ci.com/task/5415734172450816?logs=clone#L23) for an example. It shows:
```
Automatic merge went well; stopped before committing as requested
```
and the job fails further down with an issue that's fixed already in `main`. The problem stems from the `--no-commit` here: https://github.com/scipy/scipy/blob/f607e5df463bd164765769342d4659111d52a86f/ci/cirrus_general_ci.yml#L23-L26

@andyfaff glancing at `git blame`, I'm not clear on what happened there - maybe it got broken ~8 months ago in the Boost PR that touched those lines after your last change to deal with Alpine. Would you be able to have a look at that?

**Comment by andyfaff** [comment]
The cirrus clone has the following logic:

- git clone the main https://github.com/scipy/scipy.git repo.
- fetch the PR changeset (git fetch origin pull/$CIRRUS_PR/head:pull/$CIRRUS_PR)
- checkout the main branch of the repo, this should be fully up to date
- merge in the PR changeset (with a `-no-commit`) on to the tip of the main branch. It shouldn't be necessary to do a commit here.
- update the submodules
- run the tests.

I tried this process on PR 19364 on my own computer and can see the change belonging to that PR in the codebase, although the commits do not show up in the log because there is no commit.

However, I've tried reproducing the run contained in the log file and am experiencing issues with merging the changeset, there is a merge conflict with the following:

```
git clone https://github.com/scipy/scipy.git /src
cd src
git fetch origin pull/19371/head:pull/19371
git checkout main
git -c user.email="you@example.com" merge --no-commit pull/19371
```

So I don't know how the log says that the automatic merge worked.  I also tried:

```
git fetch origin
git checkout main
git rebase origin/main
git checkout pull/19371
git rebase origin/main
```
The rebase step also produces a merge conflict. I don't know how Github CI manages the clone and merge.

**Comment by andyfaff** [comment]
@rgommers, can you try merging the changeset in 19371 on your local computer?

**Comment by rgommers** [comment]
Yes, it has a conflict, as the GitHub UI also indicates in gh-19731:
```
$ gh pr checkout 19371  # creates `ansari_anp` branch and switching to it
$ git checkout main
$ git merge ansari_anp 
Auto-merging scipy/stats/tests/test_axis_nan_policy.py
CONFLICT (content): Merge conflict in scipy/stats/tests/test_axis_nan_policy.py
Automatic merge failed; fix conflicts and then commit the result.
```

`--no-commit` makes no difference here, either way there's a conflict marker in the file:
```diff
<<<<<<< HEAD
    (stats.entropy, tuple(), dict(), 1, 1, False, lambda x: (x,)),
    (stats.entropy, tuple(), dict(), 2, 1, True, lambda x: (x,))
=======
    (stats.ansari, tuple(), {}, 2, 2, False, None),
>>>>>>> ansari_anp
]
```
If that happens, the CI job should simply fail to run I think, rather than proceed and fail halfway through. How about using `git status --porcelain` here to check that (see https://stackoverflow.com/questions/5139290/how-to-check-if-theres-nothing-to-be-committed-in-the-current-branch)?

**Comment by andyfaff** [comment]
I'm surprised that the failed merge didn't stop the run

**Comment by dschmitz89** [comment]
Are we still using Cirrus CI or planning to use it somewhere? If not, we can close this.
